matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/network/parsing/CredentialsResponseAdapter.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2022 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.matrix.android.sdk.internal.network.parsing
+
+import com.squareup.moshi.FromJson
+import org.matrix.android.sdk.api.auth.data.Credentials
+
+/**
+ * Used when we receive a new credential from the serer that contains an access token expiry. We want to
+ * estimate the expiry date as soon as possible using `expiresInMs`.
+ */
+class CredentialsResponseAdapter {
+    @FromJson
+    fun credentialsFromResponse(responseCredentials: Credentials): Credentials {
+        if (responseCredentials.expiresInMs == null || responseCredentials.expiryTs != null) {
+            // expiryTs already estimated or doesn't apply, return existing credential.
+            return responseCredentials
+        }
+        // We have received a credential response from the server, estimate the expiry datetime.
+        return responseCredentials.copy(expiryTs = System.currentTimeMillis() + responseCredentials.expiresInMs)
```

Comment:
I understand reading this code again that `expiryTs` does not come from the server, do it?
This a bit weird to create a customized Json adapter to handle this.
Not sure how to implement it in another way, I will think more about it.